File: tests/test_passphrase_gate.py
"""Tests for the passphrase-gate hash (tools/passphrase_gate.py).

Twin of site/js/passphrase_gate.js. SOFT gate, not encryption: the hash ships
publicly and the gated content is public on GitHub regardless. These tests pin
the PBKDF2 output and the cross-language vectors shared with the JS twin.
"""

import json
from pathlib import Path

from tools.passphrase_gate import derive_hash, main

SALT = "a3f1c92e6b4d80571e2c9f3a6d8b0e47"
ITERS = 200000
VECTORS_PATH = Path(__file__).parent / "fixtures" / "passphrase_gate_vectors.json"


def test_derive_hash_matches_known_reference():
    assert (
        derive_hash("test-passphrase", SALT, ITERS)
        == "289fa9ee16cb75d517736c68cd7d9a646fde31efead2e17b59c3219bd1548f5f"
    )


def test_derive_hash_is_64_hex_chars():
    h = derive_hash("anything", SALT, ITERS)
    assert len(h) == 64
    assert all(c in "0123456789abcdef" for c in h)


def test_cli_hash_prints_only_the_hash(capsys):
    main(["hash", "--salt", SALT, "--iterations", str(ITERS), "test-passphrase"])
    out = capsys.readouterr().out.strip()
    assert out == "289fa9ee16cb75d517736c68cd7d9a646fde31efead2e17b59c3219bd1548f5f"
    # The phrase must never leak to stdout.
    assert "test-passphrase" not in out


def test_cross_language_vectors_match():
    """Frozen vectors shared byte-for-byte with tests/test_passphrase_gate.js.

    This is the contract that keeps the Python and JS hashes identical. Phrases
    are synthetic — NEVER the real passphrase.
    """
    data = json.loads(VECTORS_PATH.read_text(encoding="utf-8"))
    assert data["salt"] == SALT
    assert data["iterations"] == ITERS
    assert data["vectors"], "vectors file must not be empty"
    for vec in data["vectors"]:
        assert derive_hash(vec["phrase"], data["salt"], data["iterations"]) == vec["hash"]

File: tests/test_passphrase_gate.js
const { describe, it } = require('node:test');
const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const Gate = require('../site/js/passphrase_gate');

const data = JSON.parse(
  fs.readFileSync(path.join(__dirname, 'fixtures', 'passphrase_gate_vectors.json'), 'utf-8')
);

function memStorage(initial) {
  const m = Object.assign({}, initial || {});
  return {
    getItem(k) { return k in m ? m[k] : null; },
    setItem(k, v) { m[k] = String(v); },
  };
}

describe('passphrase_gate — constants match the shared fixture', () => {
  it('GATE_SALT and GATE_ITERATIONS equal the fixture', () => {
    assert.strictEqual(Gate.GATE_SALT, data.salt);
    assert.strictEqual(Gate.GATE_ITERATIONS, data.iterations);
  });
});

describe('passphrase_gate — derivePassphraseHash (cross-language)', () => {
  data.vectors.forEach((vec) => {
    it('reproduces the frozen hash for ' + vec.phrase, async () => {
      assert.strictEqual(await Gate.derivePassphraseHash(vec.phrase), vec.hash);
    });
  });
});

describe('passphrase_gate — verifyPassphrase', () => {
  const expected = data.vectors[0].hash;
  it('true for the correct phrase', async () => {
    assert.strictEqual(await Gate.verifyPassphrase(data.vectors[0].phrase, expected), true);
  });
  it('false for a wrong phrase', async () => {
    assert.strictEqual(await Gate.verifyPassphrase('nope', expected), false);
  });
  it('false when the expected hash is empty (gate disabled)', async () => {
    assert.strictEqual(await Gate.verifyPassphrase(data.vectors[0].phrase, ''), false);
  });
});

describe('passphrase_gate — isGateEnabled', () => {
  it('false for empty, true for a hash', () => {
    assert.strictEqual(Gate.isGateEnabled(''), false);
    assert.strictEqual(Gate.isGateEnabled(data.vectors[0].hash), true);
  });
});

describe('passphrase_gate — unlock state (stores + compares the hash)', () => {
  const expected = data.vectors[0].hash;
  it('false when storage is empty', () => {
    assert.strictEqual(Gate.isUnlocked(expected, memStorage()), false);
  });
  it('false when the stored hash differs (rotation re-prompts)', () => {
    const s = memStorage({ sy_gate: 'old-rotated-away-hash' });
    assert.strictEqual(Gate.isUnlocked(expected, s), false);
  });
  it('true when the stored hash equals the expected hash', () => {
    const s = memStorage({ sy_gate: expected });
    assert.strictEqual(Gate.isUnlocked(expected, s), true);
  });
  it('recordUnlock writes the expected hash under sy_gate', () => {
    const s = memStorage();
    Gate.recordUnlock(expected, s);
    assert.strictEqual(s.getItem('sy_gate'), expected);
    assert.strictEqual(Gate.isUnlocked(expected, s), true);
  });
});
